MFMA + LocalSplitU(LSU) support

nakajee opened the pull request:
* added support for MFMA + LocalSplitU (only 2 or 4 can be used)
* improved the reject condition of DirectToVgpr + MIWaveGroup to cover LSU case
* added reject condition for LSU + DirectToLds
* added reject condition for LSU + StoreCInUnroll
* added test cases for MFMA + LSU in extended folder

AlexBrownAMD commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -3646,16 +3649,36 @@ def graTileAssignment(self, kernel, tP):
+      # need division for qReg
+      kStr += vectorStaticDivide(qReg, qReg, kernel["MatrixInstN"], tmpSgpr)
+
+      # localSplitU case. Calculate LSU offset here
+      if kernel["LocalSplitU"] > 1:
+          # alloc vgpr
+          tmpVgpr = self.vgprPool.checkOut(1, 'graTA vgpr', self.preventVgprOverflowDuringNewTile)
+          subGroup  = kernel["SubGroup0"] * kernel["SubGroup1"]
+          dividendForKId   = kernel["MatrixInstM"] * kernel["MatrixInstB"]
+          kInterval = waveWidth // dividendForKId
+          # generate instruction
+          kStr += vectorStaticDivide(tmpVgpr, "Serial", subGroup, tmpSgpr, \
+            "LSU offset: sgid = Serial / subGroup(%u)" % subGroup)
+          kStr += staticMultiply(vgpr(tmpVgpr), vgpr(tmpVgpr), kInterval, sgpr(tmpSgpr), \
+            "LSU offset: lsuOffset = sgid * kInterval(%u)" % (kInterval))
+          kStr += inst("_v_add_u32", vgpr(qReg), vgpr(tmpVgpr), vgpr(qReg), \
+            "LSU offset: lrOffset += lsuOffset")
+          # release register
+          self.vgprPool.checkIn(tmpVgpr)
```
Is this the same as the code in LraTileAssignment? Is the offset saved for re-use?

AlexBrownAMD commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -6509,28 +6546,29 @@ def mfmaIter(self, kernel, u, innerUnroll, vregSetIdx, lastKinloop=False, tail=F
+              # if ccA, copy ccVgprs[0] to ccVgpr[1] to avoid generating same negate value twice
+              if ccA:
+                ccVgprs[1] = ccVgprs[0]
+            elif ccA:
```
Use of elif here could use a comment, not an intuitive sequence of ifs

AlexBrownAMD replied:
or something like

if cca==ccb
else
  if cca
  elif ccb

might be a bit easier to read

nakajee replied:
> or something like
> 
> if cca==ccb else if cca elif ccb
> 
> might be a bit easier to read

elif ccb is not correct.
We need ccb code regardless of cca.
cca==ccb part and cca part essentially generates the same code.
We would like to avoid generating same code twice.
I can add some more comment here.

nakajee replied:
How about the comment like this?

================
            # negate code for Ai
            # necessary when ccA == ccB or ccA
            # the location to add negation is different between ccA == ccB and ccA
            #   index = 0 for ccA == ccB
            #   index = 1 for ccA
            # if ccA == ccB and ccA, we need negation twice for index 0 and 1 (but we need to calculate -Ai only once)
            if ccA == ccB:
===================

nakajee replied:
Updated. I will start another PR for extra optimizations.

AlexBrownAMD reviewed (commented):
This is a big PR, There's a few unrelated changes here that could have been separate PRs (better for reviewing, testing, debugging, error tracking).
This will change some aspects of already tuned kernels. Have you tried running rocblas tests with these changes?

nakajee commented on Tensile/KernelWriter.py:
```diff
@@ -3333,6 +3333,13 @@ def kernelBody( self, kernel, tensorParametersA, tensorParametersB ):
         kl.append(self.comment("LocalSplitU: local write"))
         kl.append(self.localSplitULocalWrite(kernel))
 
+        # Use non-MI pass for MI + LSU>1
+        backup = kernel["EnableMatrixInstruction"]
+        if backup:
+          serializedStoreBackup = self.serializedStore
+          kernel["EnableMatrixInstruction"] = False
```
> Temporarily changing the kernel parameter off and on seems like it could be error prone, or confusing when debugging. If some of the localSplitU functions are dependent on EnableMatrixInstruction could we instead make a separate internal parameter that defiunes this parameter combo that these functions could check?

Global store code is huge and it is very difficult to go through non-MatrixInstruction code with keeping EnableMatrixInstruction = true.
I thought adding a lot of changes to force MI+LSU to go through non-MFMA path could cause unexpected fails...

nakajee replied:
> > Temporarily changing the kernel parameter off and on seems like it could be error prone, or confusing when debugging. If some of the localSplitU functions are dependent on EnableMatrixInstruction could we instead make a separate internal parameter that defiunes this parameter combo that these functions could check?
> 
> Global store code is huge and it is very difficult to go through non-MatrixInstruction code with keeping EnableMatrixInstruction = true. I thought adding a lot of changes to force MI+LSU to go through non-MFMA path could cause unexpected fails...

I will check the code to see if I can create a new parameter like EnableMatrixInstructionStore or something like that.
Plus, probably I can set self.serializedStore=False at the beginning for MFMA+LSU and no backup needed for self.serializedStore.

nakajee replied:
Updated

nakajee commented:
> This is a big PR, There's a few unrelated changes here that could have been separate PRs (better for reviewing, testing, debugging, error tracking). This will change some aspects of already tuned kernels. Have you tried running rocblas tests with these changes?

I have not tried rocblas yet.
These optimizations (except for MFMA+LSU) are basically removing redundant instructions.
From performance point of view, I do not expect it makes performance worse.

Additional optimizations are not much big. If necessary, I can create a separate PR for additional optimization part.

nakajee commented:
> > This is a big PR, There's a few unrelated changes here that could have been separate PRs (better for reviewing, testing, debugging, error tracking). This will change some aspects of already tuned kernels. Have you tried running rocblas tests with these changes?
> 
> I have not tried rocblas yet. These optimizations (except for MFMA+LSU) are basically removing redundant instructions. From performance point of view, I do not expect it makes performance worse.
> 
> Additional optimizations are not much big. If necessary, I can create a separate PR for additional optimization part.

Updated. I removed extra optimization.

nakajee commented:
The gfx90a extended failed last time (number 2).
I re-ran it and it passed this time (number 3).
I tried to reproduce the fail on my local node many times, but it all passed. 
The fail is not reproducible.
